File: domain/src/main/java/com/arturogutierrez/openticator/domain/issuer/interactor/GetIssuersInteractor.java
package com.arturogutierrez.openticator.domain.issuer.interactor;

import com.arturogutierrez.openticator.domain.issuer.model.Issuer;
import com.arturogutierrez.openticator.domain.issuer.repository.IssuerRepository;
import com.arturogutierrez.openticator.executor.PostExecutionThread;
import com.arturogutierrez.openticator.executor.ThreadExecutor;
import com.arturogutierrez.openticator.interactor.Interactor;
import java.util.List;
import rx.Observable;

public class GetIssuersInteractor extends Interactor<List<Issuer>> {

  private final IssuerRepository issuerRepository;

  public GetIssuersInteractor(IssuerRepository issuerRepository, ThreadExecutor threadExecutor,
      PostExecutionThread postExecutionThread) {
    super(threadExecutor, postExecutionThread);
    this.issuerRepository = issuerRepository;
  }

  @Override
  public Observable<List<Issuer>> createObservable() {
    return issuerRepository.getIssuers();
  }
}

File: domain/src/main/kotlin/com/arturogutierrez/openticator/domain/issuer/interactor/GetIssuersInteractor.kt
package com.arturogutierrez.openticator.domain.issuer.interactor

import com.arturogutierrez.openticator.domain.issuer.interactor.GetIssuersInteractor.EmptyParams
import com.arturogutierrez.openticator.domain.issuer.model.Issuer
import com.arturogutierrez.openticator.domain.issuer.repository.IssuerRepository
import com.arturogutierrez.openticator.executor.PostExecutionThread
import com.arturogutierrez.openticator.executor.ThreadExecutor
import com.arturogutierrez.openticator.interactor.SingleUseCase
import io.reactivex.Single

class GetIssuersInteractor(private val issuerRepository: IssuerRepository,
                           threadExecutor: ThreadExecutor,
                           postExecutionThread: PostExecutionThread)
  : SingleUseCase<List<Issuer>, EmptyParams>(threadExecutor, postExecutionThread) {

  override fun buildObservable(params: EmptyParams): Single<List<Issuer>> {
    return issuerRepository.issuers
  }

  object EmptyParams
}
